src/engine/strat_engine/backstore/blockdev/v2.rs

```diff
@@ -0,0 +1,467 @@
+    }
+
+    /// Allocate room for RAID metadata from the back of the device.
+    pub fn alloc_raid_meta(&mut self, size: Sectors) {
+        let segs = self.used.alloc_front(size);
+        for (start, len) in segs.iter() {
+            self.raid_meta_allocs.push((*start, *len));
+        }
+    }
+
+    /// Allocate room for integrity metadata from the back of the device.
+    pub fn alloc_int_meta_back(&mut self, size: Sectors) {
+        let segs = self.used.alloc_back(size);
+        for (start, len) in segs.iter() {
+            self.integrity_meta_allocs.push((*start, *len));
+        }
+    }
+
+    /// Set the newly detected size of a block device.
+    pub fn set_new_size(&mut self, new_size: Sectors) {
+        match self.bda.dev_size().cmp(&BlockdevSize::new(new_size)) {
+            Ordering::Greater => {
+                warn!(
+                    "The given device with path: {}, UUID; {} appears to have shrunk; you may experience data loss",
+                    self.devnode().display(),
+                    self.bda.dev_uuid(),
+                );
+                self.new_size = Some(new_size);
+            }
+            Ordering::Less => {
+                self.new_size = Some(new_size);
+            }
+            Ordering::Equal => {
+                self.new_size = None;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    pub fn invariant(&self) {
+        assert!(self.total_size() == self.used.size());
+    }
+}
+
+impl InternalBlockDev for StratBlockDev {
+    fn uuid(&self) -> DevUuid {
+        self.bda.dev_uuid()
+    }
+
+    fn device(&self) -> &Device {
+        &self.dev
+    }
+
+    fn physical_path(&self) -> &Path {
+        &self.devnode
+    }
+
+    fn blksizes(&self) -> StratSectorSizes {
+        self.blksizes
+    }
+
+    fn metadata_version(&self) -> StratSigblockVersion {
+        self.bda.sigblock_version()
+    }
+
+    fn total_size(&self) -> BlockdevSize {
+        self.bda.dev_size()
+    }
+
+    fn available(&self) -> Sectors {
+        self.used.available()
+    }
+
+    fn metadata_size(&self) -> Sectors {
+        self.bda.extended_size().sectors()
+            + self.integrity_meta_allocs.iter().map(|(_, len)| *len).sum()
+            + self.raid_meta_allocs.iter().map(|(_, len)| *len).sum()
+    }
+
+    fn max_stratis_metadata_size(&self) -> MDADataSize {
+        self.bda.max_data_size()
+    }
+
+    fn in_use(&self) -> bool {
+        self.used.used() > self.metadata_size()
+    }
+
+    fn alloc(&mut self, size: Sectors) -> PerDevSegments {
+        self.used.alloc_front(size)
+    }
+
+    fn calc_new_size(&self) -> StratisResult<Option<Sectors>> {
+        let s = Self::scan_blkdev_size(self.devnode())?;
+        if Some(s) == self.new_size
+            || (self.new_size.is_none() && s == self.bda.dev_size().sectors())
+        {
+            Ok(None)
+        } else {
+            Ok(Some(s))
+        }
+    }
+
+    fn grow(&mut self) -> StratisResult<bool> {
```

Comment:
I _believe_ that the implementation of this method should also include an adjustment to the ```integrity_meta_allocs``` value since the amount of metadata space required increases with the size of the device.